File: runtime/flim.lua
function each(fn, ...)
  for _, v in ipairs(...) do
    fn(v)
  end
end

function eachn(fn, ...)
  local arrays = {...}
  for i = 1, # arrays[1] do
    fn(unpack(map(function(arr)
      return arr[i]
    end, ...)))
  end
end

function map(fn, ...)
  local results = {}
  local array = {...}
  for i,v in ipairs(array) do
    results[i] = fn(v)
  end
  return results
end

function mapn(fn, ...)
  local results = {}
  local arrays = {...}
  for i = 1, # arrays[1] do
    results[i] = fn(unpack(map(function(arr)
      return arr[i]
    end, ...)))
  end
  return results
end

function pick(a, b, condition)
  if condition then return a else return b end
end

function chord(time, pitches, velocity, duration, channel)
  for _, p in ipairs(pitches) do
    note(time, p, velocity, duration, channel)
  end
end

function make_metro(tempo)
  local tempo = tempo
  local time = now()
  local g_tempo = 60.0 / tempo

  return function(str, ...)
    if type(str) == "number" then
      return str * g_tempo * 1000
    elseif str == 'beat' then
      return math.ceil(((now() - time) / (1000 * g_tempo)))
    end
  end
end

function array_to_str(array)
  local str = "["
  for _, v in ipairs(array) do
    str = str .. tostring(v)
  end
  str = str .. "]"
  return str
end

function sum(array)
  local result = 0
  for _,v in ipairs(array) do
    result = result + v
  end
  return result
end

function rest(array)
  local dup = clone(array)
  table.remove(dup, 1)
  return dup
end

function clone(array)
  local c = {}
  for _, value in ipairs(array) do
    table.insert(c, value)
  end
  return c
end

function make_metre(metre, base)
  local metre_length = sum(metre)
  return function(time, beat)
    local qtime = (time / base) % metre_length
    local metre = metre
    local a = metre[1]
    local b = 0
    while (qtime >= a) do
      a = a + metre[2]
      b = b + metre[1]
      metre = rest(metre)
    end
    return 1.0 + (qtime - b) == beat
  end
end

File: runtime/flim.js
this.cos = Math.cos
this.random = Math.random

function metronome(bpm) {
  var bpm = bpm,
      beatLength = (60.0 / bpm) * 1000.0,
      start = now()

  return {
    time: function(numberOfBeats) {
      return numberOfBeats * beatLength
    },

    beat: function() {
      return Math.ceil(now() / beatLength)
    },

    dur: function() {
      return beatLength
    }

  }
}

function makeMetre(numerators, denominator) {
  var metreLength = _.inject(numerators, function(sum, n) {
    return sum += n
  }, 0)

  return function(time, beat) {
    var qtime = (time / denominator) % metreLength,
        metre = numerators,
        a = metre[0],
        b = 0

    while (qtime >= a) {
      a = a + metre[1]
      b = a + metre[0]
      metre = _.rest(metre)
    }
    return 1.0 + (qtime - b) == beat
  }
}

function rand(min, max) {
  return Math.round(Math.random() * (max - min) + min)
}

function generator() {
  var args = Array.slice(arguments)
  var index = 0
  return function() {
    if (index == args.length) index = 0
    return args[index++]
  }
}

function instrument(options) {
  var options = options || {}
  var params = _.defaults(options, {
    pitch: 60,
    velocity: 80,
    duration: 100,
    channel: 1
  })
  return {
    play: function(time) {
      var args = Array.slice(arguments)
      var pitch    = args[1] || params.pitch
      var velocity = args[2] || params.velocity
      var duration = args[3] || params.duration
      var channel  = args[4] || params.channel
      note(time, pitch, velocity, duration, channel)
    }
  }
}

function playChord(time, pitches, velocity, duration, channel) {
  _.each(pitches, function(pitch) {
    note(time, pitch, velocity, duration, channel)
  })
}

function assoc(object, lists) {
  return first(select(lists, function(list) {
    return equals(first(list), object)
  }))
}

function sample(list) {
  return list[rand(0, list.length - 1)]
}
